sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -313,27 +315,24 @@ public FileBasedWriteOperation(FileBasedSink<T> sink) {
           sink.getBaseOutputFilenameProvider(), new TemporaryDirectoryBuilder()));
     }
 
-    private static class TemporaryDirectoryBuilder
+    @VisibleForTesting
+    static class TemporaryDirectoryBuilder
         implements SerializableFunction<String, String> {
       // The intent of the code is to have a consistent value of tempDirectory across
       // all workers, which wouldn't happen if now() was called inline.
       Instant now = Instant.now();
 
       @Override
       public String apply(String baseOutputFilename) {
-        try {
-          IOChannelFactory factory = IOChannelUtils.getFactory(baseOutputFilename);
-          Path baseOutputPath = factory.toPath(baseOutputFilename);
-          return baseOutputPath
-              .resolveSibling(
-                  "temp-beam-"
-                  + baseOutputPath.getFileName()
-                  + "-"
-                  + now.toString(DateTimeFormat.forPattern("yyyy-MM-DD_HH-mm-ss")))
-              .toString();
-        } catch (IOException e) {
-          throw new RuntimeException(e);
-        }
+        URI baseOutputUri = URI.create(baseOutputFilename).normalize();
```

Comment:
1 - ok, I'm convinced.

2. "non-URI paths need to be converted" - which code does that? In other words: does the current PR change user-facing behavior of FileBasedSink in a backward-incompatible way on Windows, i.e. will it break existing user pipelines? If yes, please fix that; if no, please explain why not, because if I'm understanding correctly the answer currently is yes.

On bullets 2,3 - please put them in comments in code - I asked about them because they are unclear from the code; and ideally they should be.